File: tests/typesafety/python/escape_wrapper_misuse.py
# expect: fails
# title: A raw note into render_note, and a second escape, both rejected by the checker
# misuse-of: escape_wrapper
from dataclasses import dataclass


@dataclass(frozen=True)
class HtmlText:
    value: str


def escape_html(raw: str) -> HtmlText:
    escaped = raw.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    return HtmlText(escaped)


def render_note(note: HtmlText) -> str:
    return f"<p>{note.value}</p>"


page = render_note("<b>fragile</b>")
double = escape_html(escape_html("<b>fragile</b>"))

File: tests/typesafety/typescript/escape_wrapper_misuse.ts
// expect: fails

type HtmlText = { readonly html: string };

function escapeHtml(raw: string): HtmlText {
  return {
    html: raw.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;"),
  };
}

function renderNote(note: HtmlText): string {
  return `<p>${note.html}</p>`;
}

export const page = renderNote("<b>fragile</b>");
export const twice = escapeHtml(escapeHtml("<b>fragile</b>"));
